articles/tutorials/advanced/2d_shaders/06_color_swap_effect/index.md

```diff
@@ -0,0 +1,643 @@
+
+To help visualize the effect, it will be helpful to visualize the original color _and_ the swap color. Add a control parameter that can be used to select between the two colors:
+
+```hlsl
+// a control variable to lerp between original color and swapped color  
+float OriginalAmount;
+```
+
+Change the shader function to the following:
+
+```hlsl
+float4 MainPS(VertexShaderOutput input) : COLOR
+{
+    // read the original color value
+    float4 originalColor = tex2D(SpriteTextureSampler,input.TextureCoordinates);
+    
+    // produce the key location
+    //  note the x-offset by half a texel solves rounding errors.
+    float2 keyUv = float2(originalColor.r;
+    
+    // read the swap color value
+    float4 swappedColor = tex2D(ColorMapSampler, keyUv) * originalColor.a;
+    
+    // return the result color
+    return lerp(swappedColor, originalColor, OriginalAmount);
+}
+```
+
+Now in the game, we can visualize the color swap by adjusting the control parameter. Perhaps the colors we picked do not look very nice.
+![Figure 6.8: The color swap effect is working!](./gifs/color-swap.gif)
+
+That looks pretty good, but changing between original and swap colors reveals a visual glitch. The color table didn't account for _some_ of the original colors. All of the colors get mapped, and our default color in the map was _white_, so some of the game's art is just turning white. For example, look at the torches on the top-wall. 
+
+To fix this, we can adjust the color lookup map to use transparent values by default. Use [this texture](./images/color-map-2.png) instead.
+![Figure 6.9: The color map with a transparent default color](./images/color-map-2.png)
+
+Now, anytime the swapped color value has an `alpha` value of zero, the implication is that the color was not part of the table. In that case, the shader should default to the original color instead of the non-existent mapped value.
+
+In the shader, before the final `return` line, add this snippet:
+
+```hlsl
+// ignore the swap if the map does not have a value  
+bool hasSwapColor = swappedColor.a > 0;  
+if (!hasSwapColor)  
+{  
+    return originalColor;  
+}
+```
+
+![Figure 6.10: Colors that are not in the map do not change color](./gifs/color-swap-2.gif)
+
+One final glitch becomes apparent if you stare at that long enough, which is that the center pixel in the torch is changing color from its original _white_, to our mapped orange color. In a way, that is _by design_, because the white values are being mapped. Fixing this would require a modification to the original assets to change the color the torch center, but that is left as an exercise for the reader. 
+
+### Nicer Colors
+
+The colors used above aren't the nicest. They were used for demonstration purposes. Here are some nicer textures to use that produce better results. 
+
+Dark Purple - Here is the color map for a [dark-purple](./images/color-map-dark-purple.png) color scheme.
```

Comment:
Rather than taking up so much space and making it easier to compare, this would probably look bettter as a 2x2 grid, like here for example:

https://github.com/cdhanna/docs.monogame.github.io/blob/2aef6d967b5ed7ecc66816177f1532059b37ed24/articles/tutorials/building_2d_games/01_what_is_monogame/index.md?plain=1#L20